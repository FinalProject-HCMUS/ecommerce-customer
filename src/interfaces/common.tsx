export interface IRoute {
    exact?: boolean;
    path: string;
    name: string;
    component?: React.ElementType;
    children?: string[];
    flagKey?: string;
    root?: string;
}