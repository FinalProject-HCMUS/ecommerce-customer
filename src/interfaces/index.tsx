export * from './blog/BlogResponse'
export * from './category/CategoryResponse'
export * from './message/MessageResponse'
export * from './order/OrderDetailResponse'
export * from './order/OrderResponse'
export * from './order/OrderTrackResponse'
export * from './order/PaymentMethod'
export * from './order/Status'
export * from './product/CartItemResponse'
export * from './product/ColorResponse'
export * from './product/ProductColorSizeResponse'
export * from './product/ProductImageResponse'
export * from './product/ProductResponse'
export * from './product/TopProduct'
export * from './product/SizeResponse'
export * from './review/ReviewResponse'
export * from './user/UserResponse'
export * from './user/CreateUserRequest'
export * from './user/Role'
export * from './region/region'
export * from './common/CustomResponse'
export * from './common/Pageable'
export * from './common/CustomError'
