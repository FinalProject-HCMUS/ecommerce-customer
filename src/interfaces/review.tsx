export interface Review {
  id: number;
  author: string;
  isVerified: boolean;
  rating: number;
  content: string;
  date: string;
}
