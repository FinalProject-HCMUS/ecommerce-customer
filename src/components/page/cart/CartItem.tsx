import React from 'react';
import { Trash2 } from 'lucide-react';
import QuantityControl from './QuantityControl';
import { CartItemResponse } from '../../../interfaces';
import { t } from '../../../helpers/i18n';
import { Link } from 'react-router-dom';

export interface CartItemProps {
  item: CartItemResponse;
  updateQuantity: (id: string, quantity: number) => void;
  removeItem: (id: string) => void;
  isSelected: boolean;
  onSelect: (id: string, isSelected: boolean) => void;
}

const CartItem: React.FC<CartItemProps> = ({
  item,
  updateQuantity,
  removeItem,
  isSelected,
  onSelect,
}) => {
  return (
    <div className="bg-white rounded-[12px] p-6 border border-gray-200">
      <div className="flex items-center">
        <div className="flex items-center mr-4">
          <input
            type="checkbox"
            id={`item-${item.id}`}
            checked={isSelected}
            onChange={(e) => onSelect(item.id, e.target.checked)}
            className="h-5 w-5 text-blue-600 rounded focus:ring-blue-500"
            aria-label={`Select ${item.product.name} for checkout`}
          />
        </div>

        <div className="h-20 w-20 flex-shrink-0 overflow-hidden rounded-[12px] border border-gray-200">
          <img
            src={item.product.mainImageUrl}
            className="h-full w-full object-cover object-center"
            alt={item.product.name}
          />
        </div>

        <div className="ml-4 flex-1">
          <div className="flex justify-between">
            <div>
              <Link
                to={`/product/${item.product.id}`}
                className="text-lg font-medium text-gray-900 hover:text-blue-600"
              >
                {item.product.name}
              </Link>
              <p className="mt-1 text-sm text-gray-500">{`${t('lbl.size')}: ${item.size}`}</p>
              <p className="mt-1 text-sm text-gray-500">
                {t('lbl.color')}:{' '}
                <button
                  style={{
                    backgroundColor: item.color,
                  }}
                  className="w-3 h-3 rounded-full transition-transform duration-200 hover:scale-105 py-1"
                />
              </p>
              <p className="mt-2 text-md font-medium text-gray-900">
                {(
                  (item.product.price * (100 - item.product.discountPercent)) /
                  100
                ).toLocaleString('vi-VN')}{' '}
                â‚«
              </p>
            </div>
            <div className="flex flex-col items-end">
              <button
                onClick={() => removeItem(item.id)}
                className="text-red-500 hover:text-red-700"
                data-testid={`remove-item-${item.id}`}
              >
                <Trash2 size={18} />
              </button>
              <div className="flex flex-end mt-4">
                <QuantityControl
                  quantity={item.quantity}
                  onDecrease={() => updateQuantity(item.id, item.quantity - 1)}
                  onIncrease={() => updateQuantity(item.id, item.quantity + 1)}
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

export default CartItem;
