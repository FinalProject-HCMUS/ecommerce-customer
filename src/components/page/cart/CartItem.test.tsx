import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { BrowserRouter } from 'react-router-dom';
import CartItem from './CartItem';
import { CartItemResponse } from '../../../interfaces';

// Mock dependencies
jest.mock('../../../helpers/i18n', () => ({
  t: (key: string) => key,
}));

describe('CartItem', () => {
  // Mock props with complete interface matching
  const mockItem: CartItemResponse = {
    id: 'cart-item-1',
    quantity: 2,
    product: {
      id: 'product-1',
      name: 'Test Product',
      description: 'Test description',
      cost: 80000,
      total: 200000,
      price: 100000,
      discountPercent: 10,
      enable: true,
      inStock: true,
      mainImageUrl: '/test-image.jpg',
      averageRating: 4.5,
      reviewCount: 12,
      categoryId: 'category-1',
      createdAt: '2023-01-01',
      updatedAt: '2023-01-01',
      createdBy: 'admin',
      updatedBy: 'admin',
    },
    color: '#FF0000',
    size: 'M',
    userId: 'user-1',
    itemId: 'item-1',
  };

  const mockProps = {
    item: mockItem,
    updateQuantity: jest.fn(),
    removeItem: jest.fn(),
    isSelected: false,
    onSelect: jest.fn(),
  };

  const renderWithRouter = (ui: React.ReactElement) => {
    return render(ui, { wrapper: BrowserRouter });
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('matches snapshot', () => {
    const { container } = renderWithRouter(<CartItem {...mockProps} />);
    expect(container).toMatchSnapshot();
  });

  test('renders the cart item with correct information', () => {
    renderWithRouter(<CartItem {...mockProps} />);

    // Check product information is displayed
    expect(screen.getByText('Test Product')).toBeInTheDocument();
    expect(screen.getByText('lbl.size: M')).toBeInTheDocument();
    expect(screen.getByText('lbl.color:')).toBeInTheDocument();

    // Check image
    const image = screen.getByAltText('Test Product');
    expect(image).toBeInTheDocument();
    expect(image).toHaveAttribute('src', '/test-image.jpg');
    // Check quantity is displayed
    expect(screen.getByText('2')).toBeInTheDocument();
  });

  test('handles checkbox selection', () => {
    renderWithRouter(<CartItem {...mockProps} />);

    const checkbox = screen.getByRole('checkbox');
    expect(checkbox).not.toBeChecked();

    fireEvent.click(checkbox);
    expect(mockProps.onSelect).toHaveBeenCalledWith('cart-item-1', true);
  });

  test('shows checked state when isSelected is true', () => {
    renderWithRouter(<CartItem {...{ ...mockProps, isSelected: true }} />);

    const checkbox = screen.getByRole('checkbox');
    expect(checkbox).toBeChecked();
  });

  test('handles remove button click', () => {
    renderWithRouter(<CartItem {...mockProps} />);

    const removeButton = screen.getByTestId('remove-item-cart-item-1');
    fireEvent.click(removeButton);

    expect(mockProps.removeItem).toHaveBeenCalledWith('cart-item-1');
  });

  test('handles quantity decrease', () => {
    renderWithRouter(<CartItem {...mockProps} />);

    const decreaseButton = screen.getByText('âˆ’');
    fireEvent.click(decreaseButton);

    expect(mockProps.updateQuantity).toHaveBeenCalledWith('cart-item-1', 1);
  });

  test('handles quantity increase', () => {
    renderWithRouter(<CartItem {...mockProps} />);

    const increaseButton = screen.getByText('+');
    fireEvent.click(increaseButton);

    expect(mockProps.updateQuantity).toHaveBeenCalledWith('cart-item-1', 3);
  });

  test('renders product link with correct URL', () => {
    renderWithRouter(<CartItem {...mockProps} />);

    const productLink = screen.getByText('Test Product');
    expect(productLink.closest('a')).toHaveAttribute(
      'href',
      '/product/product-1'
    );
  });

  test('renders with accessibility attributes', () => {
    renderWithRouter(<CartItem {...mockProps} />);

    const checkbox = screen.getByRole('checkbox');
    expect(checkbox).toHaveAttribute(
      'aria-label',
      'Select Test Product for checkout'
    );
  });
});
