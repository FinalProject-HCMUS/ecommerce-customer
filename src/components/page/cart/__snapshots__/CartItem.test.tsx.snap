// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartItem matches snapshot 1`] = `
<div>
  <div
    class="bg-white rounded-[12px] p-6 border border-gray-200"
  >
    <div
      class="flex items-center"
    >
      <div
        class="flex items-center mr-4"
      >
        <input
          aria-label="Select Test Product for checkout"
          class="h-5 w-5 text-blue-600 rounded focus:ring-blue-500"
          id="item-cart-item-1"
          type="checkbox"
        />
      </div>
      <div
        class="h-20 w-20 flex-shrink-0 overflow-hidden rounded-[12px] border border-gray-200"
      >
        <img
          alt="Test Product"
          class="h-full w-full object-cover object-center"
          src="/test-image.jpg"
        />
      </div>
      <div
        class="ml-4 flex-1"
      >
        <div
          class="flex justify-between"
        >
          <div>
            <a
              class="text-lg font-medium text-gray-900 hover:text-blue-600"
              data-discover="true"
              href="/product/product-1"
            >
              Test Product
            </a>
            <p
              class="mt-1 text-sm text-gray-500"
            >
              lbl.size: M
            </p>
            <p
              class="mt-1 text-sm text-gray-500"
            >
              lbl.color
              :
               
              <button
                class="w-3 h-3 rounded-full transition-transform duration-200 hover:scale-105 py-1"
                style="background-color: rgb(255, 0, 0);"
              />
            </p>
            <p
              class="mt-2 text-md font-medium text-gray-900"
            >
              90.000
               
              ₫
            </p>
          </div>
          <div
            class="flex flex-col items-end"
          >
            <button
              class="text-red-500 hover:text-red-700"
              data-testid="remove-item-cart-item-1"
            >
              <svg
                class="lucide lucide-trash2"
                fill="none"
                height="18"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="18"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M3 6h18"
                />
                <path
                  d="M19 6v14c0 1-1 2-2 2H7c-1 0-2-1-2-2V6"
                />
                <path
                  d="M8 6V4c0-1 1-2 2-2h4c1 0 2 1 2 2v2"
                />
                <line
                  x1="10"
                  x2="10"
                  y1="11"
                  y2="17"
                />
                <line
                  x1="14"
                  x2="14"
                  y1="11"
                  y2="17"
                />
              </svg>
            </button>
            <div
              class="flex flex-end mt-4"
            >
              <div
                class="flex items-center border border-gray-300 rounded-full"
              >
                <button
                  class="w-8 h-8 flex items-center justify-center rounded-l-full hover:bg-gray-100"
                >
                  −
                </button>
                <span
                  class="w-8 text-center"
                >
                  2
                </span>
                <button
                  class="w-8 h-8 flex items-center justify-center rounded-r-full hover:bg-gray-100"
                >
                  +
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
